charts/osm: use image digest for default images (#4108)
openservicemesh/osm: docs/release_guide.md

$ git push upstream release-<version> # ex: git push upstream release-v0.4
```

## Create and push the pre-release Git tag

> The pre-release Git tag publishes the OSM control plane images to the Openservicemesh container registry in Dockerhub, and publishes the image digests as an artificat of the pre-release Github workflow. The image digests must be used in the next step to update the default control plane image referenced in the Helm charts.

Reviewer: ```suggestion
The pre-release Git tag publishes the OSM control plane images to the Openservicemesh container registry in Dockerhub, and publishes the image digests as an artifact of the pre-release Github workflow. The image digests must be used in the next step to update the default control plane image referenced in the Helm charts.
```

The pre-release Git tag is of the form `pre-rel-<release-version>`, e.g. `pre-rel-v0.4.0`.

```console
$ PRE_RELEASE_VERSION=<pre-release-version> # ex: PRE_RELEASE_VERSION=pre-rel-v0.4.0
$ git tag "$PRE_RELEASE_VERSION"
$ git push upstream "$PRE_RELEASE_VERSION"
```

> Once the pre-release Git tag has been pushed, wait for the Pre-release Github workflow to complete. Upon workflow completion, retrieve the image digests for the given release. The image digests are uploaded as a part of the Pre-release workflow. Note that multiple image digest files will be present in the uploaded artificat, one per image tag, e.g. v0.4.0 and latest. Use the release version specific image digest file to extract the image digests to be used as the default for the control plane images, i.e. osm_image_digest_v0.4.0.txt instead of osm_image_digest_latest.txt if the release version if v0.4.0.

Reviewer: ```suggestion
Once the pre-release Git tag has been pushed, wait for the Pre-release Github workflow to complete. Upon workflow completion, retrieve the image digests for the given release. The image digests are uploaded as a part of the Pre-release workflow. Note that multiple image digest files will be present in the uploaded artifact, one per image tag, e.g. v0.4.0 and latest. Use the release version specific image digest file to extract the image digests to be used as the default for the control plane images, i.e. osm_image_digest_v0.4.0.txt instead of osm_image_digest_latest.txt if the release version is v0.4.0.
```

Reviewer: For context, Shashank and I do think this flow of build images, grab digests, update references to the digests , do the rest of the release could all be automated in a single release CI workflow in the future. Having some manual intervention is the easiest thing to do right now to unblock the rest of these changes and allows us the chance to work around any issues we may run into with this process.

Author: While I think it would be ideal to have a single workflow, I don't think it's possible because there are code changes necessary once the image digests are available. We might need to auto-commit a PR to achieve that or manipulate tags in a hacky way. I don't have any ideas to coalesce these 2 steps right now.

Create a new commit on the patch branch to update the hardcoded version information in the following locations:

> * The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)

Author: @nojnhuh That's a valid question I didn't address in my commit message, but I think the release workflow should be updated to:
1. Publish the images with the appropriate release tag
2. Make the changes to retrieve the digest from the published images as a part of building the binaries for the release

Is that something that is feasible?

> * The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)

Reviewer: > 1. Publish the images with the appropriate release tag

So we move image builds from the CI environment to local workstations? I don't see that being a major problem, but does open up opportunities to make mistakes.

> 2. Make the changes to retrieve the digest from the published images as a part of building the binaries for the release

Then where do you verify the retrieved digests are the same as the digests of the images you built?

> Is that something that is feasible?

Probably? I'm not totally sure. I'd like to see what it would take to get fully reproducible builds, since if we could leverage that then it wouldn't take as much tweaking to the overall process.

> * The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)

Author: > > 1. Publish the images with the appropriate release tag
> 
> So we move image builds from the CI environment to local workstations? I don't see that being a major problem, but does open up opportunities to make mistakes.
> 
No, we could make it a part of the CI but it could be a pre-release step, where we make a code change to `constants.go` to update the image references and publish the images with the appropriate tags. 

> > 1. Make the changes to retrieve the digest from the published images as a part of building the binaries for the release
> 
> Then where do you verify the retrieved digests are the same as the digests of the images you built?
> 
This can be retrieved from the image manifest. We wouldn't need to verify anything else other than the fact that the digest matches the image that was published. This is much like updating the tags currently, with correctness of the update verified through code reviews.

> > Is that something that is feasible?
> 
> Probably? I'm not totally sure. I'd like to see what it would take to get fully reproducible builds, since if we could leverage that then it wouldn't take as much tweaking to the overall process.

What's the use case for reproducible builds? I am not sure how using image digests could play a role here. Could you share an example?

> * The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)

Reviewer: > No, we could make it a part of the CI but it could be a pre-release step, where we make a code change to `constants.go` to update the image references and publish the images with the appropriate tags.

There's still the nuance where the init image needs to be built before the injector image so the injector know the init image's digest. That's certainly possible to configure, just might be a bit convoluted. What would the trigger for this CI job be?

> This can be retrieved from the image manifest. We wouldn't need to verify anything else other than the fact that the digest matches the image that was published. This is much like updating the tags currently, with correctness of the update verified through code reviews.

I'm not exactly sure which digests you're suggesting to compare. I was trying to say we should make sure the digests we put in values.yaml and elsewhere come straight from the images built in CI and not after some push and pull if we're trying to protect against images that have been modified in Docker Hub.

> What's the use case for reproducible builds? I am not sure how using image digests could play a role here. Could you share an example?

If builds are reproducible, then we can build the images locally, see what the digests are, update image references accordingly like we do today with tags, then tag and push that commit. The CI would then build and publish images that result in the same digests since it's working off of the same tagged commit.

> * The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)

Author: @nojnhuh the digest is calculated on push to a registry. So we will not be able to calculate this beforehand. The goal is to ensure the digest is correct and bake that into the chart values **before** cutting a release.

> * The container image digests `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml)

Reviewer: > @nojnhuh the digest is calculated on push to a registry. So we will not be able to calculate this beforehand.

The digest is a property of the image itself, so we only need to build it to know what the digest is. No push is necessary. `docker inspect <image> -f '{{ .ID }}'`

> The goal is to ensure the digest is correct and bake that into the chart values before cutting a release.

Yeah I think we're agreeing here. Let's make sure all the CI changes for release are captured in this PR. Commenting on code is much easier than English 😄

Create a new commit on the patch branch to update the hardcoded version information in the following locations:

* The control plane image digests  defined by `OpenServiceMesh.image.digest` for images in [charts/osm/values.yaml](/charts/osm/values.yaml) from the image digests obtained from the Pre-release workflow. For example, if the osm-controller image digest is `sha256:eb194138abddbe271d42b290489917168a6a891a3dabb575de02c53f13879bee`, update the value of `OpenServiceMesh.image.digest.osmController` to `sha256:eb194138abddbe271d42b290489917168a6a891a3dabb575de02c53f13879bee`.

> * Replace the `latest-main` tag with the release version tag for all images throughout the repo, e.g. `v0.4.0`.

Reviewer: Does this mean the places that were listed here before, like docs/example/manifests/apps/bookbuyer.yaml? I know it was really verbose and needed to be changed fairly often, but I appreciated it when it came time to make a release since it's not as easy as a global find/replace across the repo.

Even if we don't want a super verbose list, even mentioning the general locations where these tags exist would be helpful, like docs/example/manifests.

Author: Added a note regarding the locations where this update is applicable.

> * Replace the `latest-main` tag with the release version tag for all images throughout the repo, e.g. `v0.4.0`. This includes updating the image image references in `charts/osm` and `docs/example/manifests` folders.

Reviewer: ```suggestion
* Replace the `latest-main` tag with the release version tag for all images throughout the repo, e.g. `v0.4.0`. This includes updating the image references in `charts/osm` and `docs/example/manifests` folders.
```

Skip this step if the release is a release candidate (RC).

> Open a pull request against the `main` branch to update the `version` field in `charts/osm/Chart.yaml` to the release version.

Reviewer: I'm guessing we can't just keep the chart version as `latest-main` because `helm` needs it to be a valid semantic version?
